### DISCUSSION TITLE
Should I index an array field in the payload and filter searches by it?

### DISCUSSION THREAD

**Discussion by diegoanvilla** [qdrant/qdrant#4816] [General]
I am seeking to implement a search functionality that filters results based on an array field in the payload. The payload would contain values such as [11145, 1158, 2458...], and the search would typically involve a single value from the array. However, I am uncertain if this approach is optimal, particularly in light of the fact that as more data is uploaded, it may lead to slower service speeds and increased resource consumption.

**Comment by timvisee** [comment]
This should be fine, as long as you create a proper payload index on the specific key that would contain the array of values.

Then you can use a simple [match](https://qdrant.tech/documentation/concepts/filtering/#match) filter.

Note that if you search without a vector (using scroll) it will not be able to use any HNSW index.

**Comment by sory25** [comment]
How could that be that payload index? since qdrant doesn't have array definitions for indexes

**Reply by timvisee** [reply]
> since qdrant doesn't have array definitions for indexes

If you have a structure like this:

```json
{ "key": [1, 2, 4, 8, 16] }
```

and you create a payload index on `"key"`, it'll index 1, 2, 4, 8 and 16 for that point.

That means that array values are supported.

**Comment by sory25** [comment]
Thanks Tim, so this works for arrays?:
PUT /collections/mycollection/index
{
    "field_name": "key",
    "field_schema": "integer"
}

**Reply by timvisee** [reply]
Correct.

**Comment by hbakri** [comment]
Hi @timvisee, thanks for the great work!

I’ve read through the documentation, and while I understand that indexes are supported for UUID lists in the payload, I’m unclear on how to dynamically update the UUIDs in the list. Specifically, I want to add or remove a UUID element from the “dataset_ids” field, but I couldn’t find a clear example or method for updating this field efficiently.

Could you please clarify the best approach to dynamically modify this list of UUIDs in the payload?

**Reply by timvisee** [reply]
You're probably looking for [set payload](https://qdrant.tech/documentation/concepts/payload/#set-payload). See the 4th example. Set the `key` to the key of the list you want to update, and simply provide the new list.

In terms of the payload indices - they are automatically updated when you change the payload data.

**Reply by hbakri** [reply]
Thanks for your really quick response! I understand the set payload approach, but what I’m specifically looking for is a way to add or remove a single UUID element from my “dataset_ids” field without having to rewrite the entire list.

Is there a way to append or remove just one element from the existing list, without having to retrieve and re-submit the entire list? I’m aiming to avoid any unnecessary overhead from re-indexing the entire list when only a small change is made.

**Reply by timvisee** [reply]
That currently is not possible. I do recommend to fetch, modify and resubmit the list of IDs.

I wouldn't worry to much about re-indexing, I don't expect it to be bad as updating payload indices is quite cheap.

**Reply by hbakri** [reply]
I completely understand, but my use case involves updating the `dataset_ids` payload field across multiple points, each with different UUID lists. The current approach—retrieving, modifying, and re-submitting the list for each point—becomes increasingly inefficient, especially when needing to add one element to N different points.

This adds up quickly, as each update requires N retrieve calls and N resubmit calls, instead of potentially having a single operation that would update all the relevant payloads in one go. This extra overhead not only impacts performance but also complicates concurrency management in my application, especially when handling a high volume of concurrent requests.

I was wondering if you might consider adding a feature in the future to append/remove a single element from the existing list for multiple points in a more efficient way?

Thanks again for your time and for considering this! Your help is greatly appreciated.
